configure.ac

```diff
@@ -54,6 +52,24 @@ AC_ARG_ENABLE(qt-dbus,
               [],
               [enable_qt_dbus=yes])
 
+# Check compiler C++11 support
+AC_LANG_PUSH(C++)
+m4_define([DETECT_CPP11_PROGRAM],
+    [AC_LANG_PROGRAM([[
+#ifndef __cplusplus
+#error "This is not a C++ compiler"
+#elif __cplusplus < 201103L
+#error "This is not a C++11 compiler"
+#endif]],
+    [[]])])
+_tmp="$CXXFLAGS"
+CXXFLAGS="$CXXFLAGS -std=c++11"
```

Comment:
IMO, the flow should be this:
We don't add any switch. We just test if it compiles with current CXXFLAGS. If it succeds it means at least c++11 mode. This would allow the current default c++14 for gcc to pass this test without forcing a downgrade to c++11. Ideally the current CXXFLAGS would be the same for libtorrent too.
If the compiler is in c++98 we just fail and suggest to the user to manually add to the CXXFLAGS the correct mode that will be the same as libtorrent.
Does this make sense?

Reply:
>IMO, the flow should be this:

I don't think this flow of checking is needed: libtorrent built with cmake will pass down the required `-std` flag properly. As for libtorrent built with autotools, well, if the compiling qbt succeeded then great; if not then users will have to **manually** make sure libtorrent & qbt are properly built with the same `-std` flag (either c++11 or c++14), i.e. the way you suggested won't provide any information about it, it cannot tell whether libtorrent is built with c++11 or c++14 or c++98, and there is no way to query it unless patching libtorrent (and they are ditching autotools)... so IMO we don't need to bother with it, we only make sure a C++11 compiler is available to use here.

Reply:
Let's say the compiler defaults to c++98 but also supports c++11 (via switch). The test will succeed here, but qbt compilation will fail.
That's why I proposed to detect if **at least** c++11 is enabled. If it isn't just fail and complain.
If it is enabled, there is still a possibility of mismatch with libtorrent, but it makes the packager responsible to use same mode. **Unless** of course libtorrent is built using cmake. Which brings in my next proposal: Wouldn't it be better to run this test after libtorrent pkg-config test? This way CXXFLAGS might be populated with the correct switches from libtorrent(cmake).

Reply:
>The test will succeed here, but qbt compilation will fail.

It means user have C++11 compiler but it does not operate in C++11 mode when building qbt.
As [proposed](https://github.com/qbittorrent/qBittorrent/pull/9835#issuecomment-437643042) by @glassez and [revised](https://github.com/qbittorrent/qBittorrent/pull/9835#issuecomment-437643582) by me, we think the real correct action is "If compiler supports c++11 and above, we should rely on mode used by libtorrent (no matter what mode is compiler default)".

>If it is enabled, there is still a possibility of mismatch with libtorrent, but it makes the packager responsible to use same mode.

Right, but why should our script try to be smart to take a blind shot and guess the correct `-std` flag? possibility of mismatch still exists! (and might be only discoverable at run time) why not just fail early up front and let the packager/user ensure the correct `-std` flag to use? As said before it is just a mere `CXXFLAGS=-std=c++11 ./configure` and also IMO having this line explicitly in build script makes the intention much clearer than relying on our smart guess.

>Which brings in my next proposal

Following my logic, I think we only need to care about our own needs, C++11 support. 
If depending on libtorrent C++ mode might lead us astray: what if user's compiler only supports c++11 and prebuilt libtorrent pkg have `-std=c++14` in CXXFLAGS? Do qbt needs it? IMO no, either user rebuild libtorrent in C++11 or upgrades his compiler.

Reply:
>prebuilt libtorrent pkg have -std=c++14 in CXXFLAGS? Do qbt needs it? 

Isn't this a yes? Otherwise, qbt sees different symbols causing runtime crashes. See issue raised here: #9485 

>It means user have C++11 compiler but it does not operate in C++11 mode when building qbt.
As proposed by @glassez and revised by me, we think the real correct action is **"If compiler supports c++11 and above,** we should rely on mode used by libtorrent (no matter what mode is compiler default)".

I am just expanding on the bold part. Check if user/system settings allow at least C++11 mode. Don't explicitly put switches for the test. If qbt compilation will eventually fail because of c++98 mode with user/system settings, then the configuration test should have failed earlier to indicate the problem.
The error message may read something like:
>You need to enable at least C++11. If libtorrent is built using C++14 then you need to enable C++14 instead. Example: CXXFLAGS=-std=c++11 ./configure